charmhub_lp_tools/reports.py

```diff
@@ -0,0 +1,207 @@
+import collections
+import json
+import logging
+import operator
+import os
+
+from datetime import datetime
+from typing import List
+try:
+    from zoneinfo import ZoneInfo
+except ImportError:
+    from backports.zoneinfo import ZoneInfo
+
+import humanize
+
+from jinja2 import Environment, FileSystemLoader, select_autoescape
+from prettytable import PrettyTable
+
+from .launchpadtools import TypeLPObject
+
+
+__THIS__ = os.path.dirname(os.path.abspath(__file__))
+NOW = datetime.now(tz=ZoneInfo("UTC"))
+
+
+class BaseBuildsReport:
+    """Base class report"""
+    def __init__(self, output: str = None):
+        """Initialize base report."""
+        self.output = output
+        self.log = logging.getLogger(f'{__name__}.{self.__class__.__name__}')
+
+    def add_build(self, charm_project, recipe, build: TypeLPObject):
+        """Add a new build object to the report.
+
+        :param charm_project: charm project the build belongs to.
+        :param build: build to add to the report.
+        """
+        raise NotImplementedError()
+
+    def generate(self):
+        """Generate report."""
+        raise NotImplementedError()
+
+
+class HtmlBuildsReport(BaseBuildsReport):
+    def __init__(self, output):
+        super().__init__(output)
+        # {'openstack': {'yoga': {'edge': [<build>, <build>,...]}}}
+        self.builds = collections.defaultdict(
+            lambda: collections.defaultdict(
+                lambda: collections.defaultdict(
+                    list
+                )
+            )
+        )
+
+    def add_build(self, charm_project, recipe, build):
+        # the same build could have been used to be released into multiple
+        # channels.
+        for channel in recipe.store_channels:
+            (track, risk) = channel.split('/')
```

Comment:
Might be worth checking that there really is a risk, and supplying 'stable' if only the track was specified?

Reply:
yeah, that would be nice for sure, I will add that.